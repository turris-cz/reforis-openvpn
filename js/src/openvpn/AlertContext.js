/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useState } from "react";
import PropTypes from "prop-types";

import { Alert } from "foris";

const AlertContext = React.createContext();

AlertContextProvider.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]),
};

export function AlertContextProvider({ children }) {
    const [alert, setAlert] = useState(null);

    return (
        <>
            {alert && <Alert type="danger" message={alert} onDismiss={() => setAlert(null)} />}
            <AlertContext.Provider value={setAlert}>
                { children }
            </AlertContext.Provider>
        </>
    );
}

export default AlertContext;
