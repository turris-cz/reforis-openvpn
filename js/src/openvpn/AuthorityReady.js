/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect, useContext } from "react";
import PropTypes from "prop-types";

import { Button, useAPIDelete, AlertContext } from "foris";

import API_URLs from "API";

AuthorityReady.propTypes = {
    serverEnabled: PropTypes.bool.isRequired,
    onReload: PropTypes.func.isRequired,
};

export default function AuthorityReady({ serverEnabled, onReload }) {
    const setAlert = useContext(AlertContext);

    const [deleteResponse, deleteCA] = useAPIDelete(API_URLs.authority);
    useEffect(() => {
        if (deleteResponse.isSuccess) {
            onReload();
        } else if (deleteResponse.isError) {
            setAlert(_("Cannot delete certificate authority"));
        }
    }, [deleteResponse, onReload, setAlert]);

    return (
        <>
            <h3>{_("Certificate authority")}</h3>
            {serverEnabled
                ? <p>{_("You can't delete the CA while the OpenVPN server is enabled. To delete the CA you need to disable the server configuration first.")}</p>
                : (
                    <>
                        <p>{_("Your certificate authority (CA) is set up properly.")}</p>
                        <Button onClick={() => deleteCA()}>{_("Delete CA")}</Button>
                    </>
                )}
        </>
    );
}
