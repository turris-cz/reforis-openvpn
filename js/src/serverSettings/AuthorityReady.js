/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import { Button, useAPIDelete, useAlert, API_STATE } from "foris";

import API_URLs from "API";

AuthorityReady.propTypes = {
    serverEnabled: PropTypes.bool.isRequired,
    onReload: PropTypes.func.isRequired,
};

export default function AuthorityReady({ serverEnabled, onReload }) {
    const [setAlert] = useAlert();

    const [deleteResponse, deleteCA] = useAPIDelete(API_URLs.authority);
    useEffect(() => {
        if (deleteResponse.state === API_STATE.SUCCESS) {
            onReload();
        } else if (deleteResponse.state === API_STATE.ERROR) {
            setAlert(_("Cannot delete certificate authority"));
        }
    }, [deleteResponse, onReload, setAlert]);

    return (
        <div className="card p-4">
            <h2>{_("Certificate authority")}</h2>
            {serverEnabled ? (
                <p>
                    {_(
                        "You can't delete the CA while the OpenVPN server is enabled. To delete the CA you need to disable the server configuration first."
                    )}
                </p>
            ) : (
                <p>
                    {_(
                        "Your certificate authority (CA) is set up properly. Please note that if you delete it all clients will have their access revoked."
                    )}
                </p>
            )}
            <div className="text-right">
                <Button
                    onClick={() => deleteCA()}
                    className="btn-danger"
                    forisFormSize
                    loading={deleteResponse.state === API_STATE.SENDING}
                    disabled={serverEnabled}
                >
                    {_("Delete CA")}
                </Button>
            </div>
        </div>
    );
}
