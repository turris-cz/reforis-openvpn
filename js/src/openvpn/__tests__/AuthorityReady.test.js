/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, getByText, fireEvent, wait } from "customTestRender";
import mockAxios from 'jest-mock-axios';

import AuthorityReady from "../AuthorityReady";
import AlertContext from "../AlertContext";

describe("<AuthorityReady />", () => {
    let componentContainer;
    const handleSuccess = jest.fn(),
        setAlert = jest.fn();

    beforeEach(() => {
        const { container } = render(
            <AlertContext.Provider value={setAlert}>
                <AuthorityReady onSuccess={handleSuccess} />
            </AlertContext.Provider>
        );
        componentContainer = container;
    });

    it("should match snapshot", () => {
        expect(componentContainer).toMatchSnapshot();
    });

    it("should send request when button is clicked", () => {
        fireEvent.click(getByText(componentContainer, "Delete CA"));
        expect(mockAxios.delete).toHaveBeenCalledTimes(1);
    });

    it("should handle error", async () => {
        fireEvent.click(getByText(componentContainer, "Delete CA"));
        mockAxios.mockError({response: {}});
        await wait(() => {
            expect(setAlert).toHaveBeenCalledWith("Cannot delete certificate authority");
        });
    });

    it("should handle success", async () => {
        fireEvent.click(getByText(componentContainer, "Delete CA"));
        mockAxios.mockResponse({});
        await wait(() => {
            expect(handleSuccess).toHaveBeenCalledTimes(1);
        });
    });
});
