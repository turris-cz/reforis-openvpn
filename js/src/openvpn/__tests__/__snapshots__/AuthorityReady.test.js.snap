// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthorityReady /> should match snapshot 1`] = `
<div>
  <h3>
    Certificate authority
  </h3>
  <p>
    Your certificate authority (CA) is set up properly.
  </p>
  <button
    class="btn btn-primary "
    type="button"
  >
     
     
    Delete CA
  </button>
</div>
`;
